sklearn/utils/_testing.py

```diff
@@ -387,6 +395,63 @@ def assert_raise_message(exceptions, message, function, *args, **kwargs):
         raise AssertionError("%s not raised by %s" % (names, function.__name__))
 
 
+def assert_allclose(
+    actual, desired, rtol=None, atol=0.0, equal_nan=True, err_msg="", verbose=True
+):
+    """
+    Adaptation of numpy.testing.assert_allclose to have tolerances
+    be set based on the input arrays' dtypes.
+
+    Parameters
+    ----------
+    actual : array_like
+        Array obtained.
+    desired : array_like
+        Array desired.
+    rtol : float, optional, default=None
+        Relative tolerance.
+        If None, it is set based on the provided arrays' dtypes.
+    atol : float, optional, default=0.
+        Absolute tolerance.
+        If None, it is set based on the provided arrays' dtypes.
+    equal_nan : bool, optional, default=True
+        If True, NaNs will compare equal.
+    err_msg : str, optional, default=''
+        The error message to be printed in case of failure.
+    verbose : bool, optional, default=True
+        If True, the conflicting values are appended to the error message.
+
+    Raises
+    ------
+    AssertionError
+        If actual and desired are not equal up to specified precision.
+
+    See Also
+    --------
+    numpy.testing.assert_allclose
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.utils._testing import assert_allclose
+    >>> x = [1e-5, 1e-3, 1e-1]
+    >>> y = np.arccos(np.cos(x))
+    >>> assert_allclose(x, y, rtol=1e-5, atol=0)
+    >>> a = np.full(shape=10, fill_value=1e-5, dtype=np.float32)
+    >>> assert_allclose(a, 1e-5)
+    """
+    dtypes = []
+
+    actual, desired = np.asanyarray(actual), np.asanyarray(desired)
+    dtypes = [actual.dtype, desired.dtype]
+
+    if rtol is None:
+        rtols = [DTYPE_RELATIVES_TOLERANCES.get(dtype, 1e-9) for dtype in dtypes]
```

Comment:
Let's keep the numpy default to `1e-7` for any other dtypes to preserve the numpy behavior for those cases.

Actually we can even simplify this to:

```suggestion
        rtols = [1e-4 if dtype == np.float32 or 1e-7 for dtype in dtypes]
```

and then remove the dict-based look-up on `DTYPE_RELATIVES_TOLERANCES`.

We can reintroduce such dict base look-up if we want to generalize this approach to `float16` support but at this point I think this is a [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).